File: py/bank.py
from money import Money


class Bank:
    def __init__(self):
        self.__rates = {}

    @staticmethod
    def __make_key(from_, to):
        return f"{from_}->{to}"

    def add_exchange_rate(self, from_, to, rate):
        key = self.__make_key(from_, to)
        self.__rates[key] = rate

    def convert(self, money, currency):
        if money.currency == currency:
            return Money(money.amount, currency)

        key = self.__make_key(money.currency, currency)
        try:
            return Money(money.amount * self.__rates[key], currency)
        except KeyError:
            raise Exception(key)

File: js/bank.js
const Money = require("./money");

class Bank {
  constructor() {
    this.exchangeRates = new Map();
  }

  makeKey(from, to) {
    return from + "->" + to
  }

  addExchangeRate(from, to, rate) {
    let key = this.makeKey(from, to);
    this.exchangeRates.set(key, rate);
  }

  convert(money, currency) {
    if (money.currency === currency) {
      return new Money(money.amount, money.currency);
    }
    let key = this.makeKey(money.currency, currency);
    let rate = this.exchangeRates.get(key);
    if (rate === undefined) {
      throw new Error(key);
    }
    return new Money(money.amount * rate, currency);
  }
}

module.exports = Bank;
